### PR TITLE
[TPU][V1][Bugfix] Fix w8a8 recompiilation with GSM8K

### PR DESCRIPTION
**Pull request by NickLucche**
This PR fixes the compilation issue we run into with `VLLM_XLA_CHECK_RECOMPILATION=1 VLLM_USE_V1=1 python -m pytest -s -v tests/tpu/test_quantization_accuracy.py` with the w8a8 llama model.
The actual output dtype may vary depending on the quantization mm we use. In this case, the hidden states are fp32. 
Hence, we pre-compile with dummy fp32 hidden states rather than assuming bf16.  

Update:

To provide more context here:  the "culprit" of the change is `torch.ops.xla.quantized_matmul` which returns fp32. Casting it down to bf16 should leverage TPU capabilities better in intermediate layers. Results:

```sh
pre
============ Serving Benchmark Result ============
Successful requests:                     1000      
Benchmark duration (s):                  171.10    
Total input tokens:                      1720000   
Total generated tokens:                  120874    
Request throughput (req/s):              5.84      
Output token throughput (tok/s):         706.46    
Total Token throughput (tok/s):          10759.14  
---------------Time to First Token----------------
Mean TTFT (ms):                          84022.40  
Median TTFT (ms):                        83897.42  
P99 TTFT (ms):                           166727.15 
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          44.75     
Median TPOT (ms):                        45.22     
P99 TPOT (ms):                           45.90     
---------------Inter-token Latency----------------
Mean ITL (ms):                           44.73     
Median ITL (ms):                         45.18     
P99 ITL (ms):                            47.79     
==================================================

post 
============ Serving Benchmark Result ============
Successful requests:                     1000      
Benchmark duration (s):                  170.58    
Total input tokens:                      1720000   
Total generated tokens:                  120508    
Request throughput (req/s):              5.86      
Output token throughput (tok/s):         706.46    
Total Token throughput (tok/s):          10789.74  
---------------Time to First Token----------------
Mean TTFT (ms):                          83928.69  
Median TTFT (ms):                        83772.92  
P99 TTFT (ms):                           166358.23 
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          44.66     
Median TPOT (ms):                        45.12     
P99 TPOT (ms):                           45.76     
---------------Inter-token Latency----------------
Mean ITL (ms):                           44.67     
Median ITL (ms):                         45.10     
P99 ITL (ms):                            46.79     
==================================================
```

### CODE REVIEW

**Review by alexm-redhat — APPROVED** [review]
Good catch!

**Review by mgoin — APPROVED** [review]

**Comment by NickLucche** [comment]
let's hold for benchmarks

File: vllm/v1/worker/tpu_model_runner.py [hunk 1]
```diff
@@ -787,7 +789,7 @@ def capture_model(self) -> None:
             num_reqs_to_sample = MIN_NUM_SEQS
             dummy_hidden = torch.randn((num_tokens, hsize),
                                        device=device,
-                                       dtype=torch.bfloat16)
+                                       dtype=self._output_dtype)
```
Comment (yaochengji) [inline vllm/v1/worker/tpu_model_runner.py:792]: Can we directly use the output of the `self.model` call, then we don't need such a dtype argument?

File: .buildkite/run-tpu-v1-test.sh [hunk 2]
```diff
@@ -28,16 +28,14 @@ docker run --privileged --net host --shm-size=16G -it \
...
     && pytest -s -v /workspace/vllm/tests/v1/tpu/test_sampler.py" \
 
 
 # TODO: This test fails because it uses RANDOM_SEED sampling
-# && VLLM_USE_V1=1 pytest -v -s /workspace/vllm/tests/tpu/test_custom_dispatcher.py \
-
-# TODO: Re-enable this after fixing recompilation in quantization.
-# && echo TEST_4 \
-# && pytest -s -v /workspace/vllm/tests/tpu/test_quantization_accuracy.py \
+# && VLLM_USE_V1=1 pytest -v -s /workspace/vllm/tests/tpu/test_custom_dispatcher.py \
```
Comment (yaochengji) [inline .buildkite/run-tpu-v1-test.sh:41]: Nit: add a new line.

**Review by yaochengji — CHANGES_REQUESTED** [review]
Thanks for fixing this!

I'm thinking about the root cause of the recompilation is due to we don't use the same code in `execute_model` and `dummy_run`. To prevent further similar bugs, can we use a common utility function which wraps all the tpu computations including encoder, `self.model`, logits processor / samper and make sure it is shared by both?

cc @robertgshaw2-redhat , then the disabling sampler logic don't need to apply to two places in your PR, https://github.com/vllm-project/vllm/pull/15662

File: .buildkite/run-tpu-v1-test.sh (hunk 2 above)
Comment (NickLucche) [reply .buildkite/run-tpu-v1-test.sh:41]: you mean at the end?

Comment (yaochengji) [reply .buildkite/run-tpu-v1-test.sh:41]: Yes

File: vllm/v1/worker/tpu_model_runner.py (hunk 1 above)
Comment (NickLucche) [reply vllm/v1/worker/tpu_model_runner.py:792]: either we pay an extra forward or the whole loop moves into dummy_run

Comment (NickLucche) [reply vllm/v1/worker/tpu_model_runner.py:792]: I was just respecting the function signature, so nothing is returned directly from dummy_run

**Comment by NickLucche** [comment]
> use the same code in execute_model and dummy_run

I am not sure, while I'd definitely like a cleaner structure, I think to match what we do in gpu model runner dummy_run only runs the decoder in mm.

Also, we're already underestimating memory as @robertgshaw2-redhat  found out in his PR.

**Comment by mgoin** [comment]
The gpu model runner has separate get_multimodal_embeddings, _dummy_run, and _dummy_sampler_run called for profiling, so I agree with Nicolo about staying true to gpu for now. I particularly don't see how it makes sense to lump mm encoder and decoder together

**Comment by yaochengji** [comment]
> The gpu model runner has separate get_multimodal_embeddings, _dummy_run, and _dummy_sampler_run called for profiling

The main difference is that TPU has to pre-compile for all the computations on device, not only the transformer backbone. But I'm fine with current situation as long as we have the recompilation check in our CI test.

**Review by yaochengji — APPROVED** [review]
LGTM, thanks for fixing this.

File: vllm/v1/worker/tpu_model_runner.py [hunk 3]
```diff
@@ -80,6 +80,7 @@ def __init__(
         self.enforce_eager = model_config.enforce_eager
         self.pin_memory = is_pin_memory_available()
         self.dtype = self.model_config.dtype
+        self._hidden_states_dtype = self.dtype
```
Comment (lsy323) [inline vllm/v1/worker/tpu_model_runner.py:83]: I think this is the same as `model_config.dtype` specifies the activation dtype, which is the same dtype as the `hidden_states_dtype` here. 

ref: https://github.com/vllm-project/vllm/blob/main/vllm/config.py#L131-L133

File: vllm/model_executor/layers/quantization/kernels/scaled_mm/xla.py [hunk 4]
```diff
@@ -97,7 +97,8 @@ def apply_weights(self,
                                              block_size=-1,
                                              int4_weight=False,
                                              quantize_activation=True)
-
+        # `quantized_matmul` output is fp32, cast it down to bf16 for perf
```
Comment (lsy323) [inline vllm/model_executor/layers/quantization/kernels/scaled_mm/xla.py:100]: nit: To be precise, `quantized_matmul` is not always generating `fp32` output, for some models/weights we generate `fp32` output because scaler is `fp32`, therefore the output from quantized matmul is promoted to `fp32`. We cast it back to `x.dtype` to ensure the activation dtype remain the same throughout the model

File: vllm/v1/worker/tpu_model_runner.py (hunk 3 above)
Comment (NickLucche) [reply vllm/v1/worker/tpu_model_runner.py:83]: I am afraid it's not as I did try running with `self.dtype` before the PR, but it would just evaluate to bf16. While the quantized_matmul, as you noted below, does and implicit cast to fp32 which was not accounted for in `model_config.dtype`

File: vllm/model_executor/layers/quantization/kernels/scaled_mm/xla.py (hunk 4 above)
Comment (NickLucche) [reply vllm/model_executor/layers/quantization/kernels/scaled_mm/xla.py:100]: absolutely!

File: vllm/v1/worker/tpu_model_runner.py (hunk 3 above)
Comment (lsy323) [reply vllm/v1/worker/tpu_model_runner.py:83]: If the model output generates a different dtype than the one specified in the config, then we have a bug in our model code. Supposely `_hidden_states_dtype` and `dtype` should always be the same
